components/role-mgt/org.wso2.carbon.identity.role.v2.mgt.core/src/main/java/org/wso2/carbon/identity/role/v2/mgt/core/dao/RoleDAOImpl.java

```diff
@@ -653,16 +653,17 @@ public void updateIdpGroupListOfRole(String roleId, List<IdpGroup> newGroupList,
     public void deleteRole(String roleId, String tenantDomain) throws IdentityRoleManagementException {
 
         String roleName = getRoleNameByID(roleId, tenantDomain);
-        if (systemRoles.contains(roleName)) {
+        if (systemRoles.contains(roleName) && !isSubOrgByTenant(tenantDomain)) {
             throw new IdentityRoleManagementClientException(OPERATION_FORBIDDEN.getCode(),
                     "Invalid operation. Role: " + roleName + " Cannot be deleted since it's a read only system role.");
         }
         int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
         UserRealm userRealm;
         try {
             userRealm = CarbonContext.getThreadLocalCarbonContext().getUserRealm();
-            if (UserCoreUtil.isEveryoneRole(roleName, userRealm.getRealmConfiguration())
-                    || isInternalAdminOrSystemRole(roleId, tenantDomain, userRealm)) {
+            if (!isSubOrgByTenant(tenantDomain) &&
```

Comment:
Let's change the order of the conditions.
isSubOrgByTenant() is IO intensive logic. Better to have it as final condition.

Reply:
Resolved by https://github.com/wso2/carbon-identity-framework/pull/5694/commits/8ebf2620b6b3b6c6108e99162e810b51d53440a3